zerver/lib/retention.py

```diff
@@ -502,6 +522,172 @@ def move_messages_to_archive(
         raise Message.DoesNotExist
 
 
+def extract_dm_group_archive_data_by_message_ids(
+    message_ids: list[int],
+    chunk_size: int = MESSAGE_BATCH_SIZE,
+) -> dict[int, DirectMessageGroupArchiveData]:
+    archive_data: dict[int, DirectMessageGroupArchiveData] = {}
+
+    # Process in chunks to avoid overwhelming the database
+    message_ids_head = message_ids
+    while message_ids_head:
+        message_ids_chunk = message_ids_head[0:chunk_size]
+        message_ids_head = message_ids_head[chunk_size:]
+
+        affected_recipients = (
+            Message.objects.select_related("recipient")
+            .filter(
+                id__in=message_ids_chunk,
+                is_channel_message=False,
+                recipient__type=Recipient.DIRECT_MESSAGE_GROUP,
```

Comment:
We can drop the `recipient__type=Recipient.DIRECT_MESSAGE_GROUP` condition and also `select_related("recipient")` when the migration to use direct message groups for 1:1 messages is complete (https://github.com/zulip/zulip/pull/35795).

This will optimize the query further.